### ISSUE TITLE
Upgrade to Elasticsearch 5

### ISSUE THREAD

**Issue by singingwolfboy** [mitodl/micromasters#2867]
Elasticsearch 5 has several changes and improvements compared to Elasticsearch 2, including [ingest pipelines](https://www.elastic.co/guide/en/elasticsearch/reference/5.2/ingest.html). Elasticsearch 2.4 will be maintained until February 2018, [according to the support matrix](https://www.elastic.co/support/eol#support-matrix). It would be better to upgrade to the latest version of Elasticsearch, in order to stay supported and to get new features in the software.

This issue depends on https://github.com/mitodl/elasticsearch-formula/pull/6. Micromasters can't upgrade to Elasticsearch 5 until the Elasticsearch formula supports it. Once that pull request is merged, we can evaluate this project for compatibility with Elasticsearch 5, and upgrade when we're ready.

**Comment by noisecapella** [comment]
It looks like Elasticsearch 5 has `wait_for_refresh` which could prevent against indexing race conditions in certain situations: https://www.elastic.co/blog/refreshing_news

**Comment by pdpinch** [comment]
Do we anticipate any blockers to this upgrade, besides mitodl/elasticsearch-formula#6 ?

Does this change our security profile at all?

**Comment by noisecapella** [comment]
I don't believe the endpoints we use would be affected by the upgrade. I investigated upgrading to Elasticsearch 5 and they had a couple of suggestions: 
 - run the Elasticsearch migration helper plugin to identify potential problems: https://github.com/elastic/elasticsearch-migration/tree/2.x I got this output:
![screenshot from 2017-03-30 17-22-37](https://cloud.githubusercontent.com/assets/863262/24526965/2b25855e-156e-11e7-8d7b-2b9b9967271b.png)

 - in that plugin you can also enable deprecation warnings. I got a few warnings relating to the search page:
   - sorting by ascending count is deprecated and will be removed in the next major version
   - size of 0 in aggregations is deprecated and will be invalid in future versions. Please specify a size greater than 0
   - Deprecated field [filter] used, replaced by [query]

**Comment by noisecapella** [comment]
I think the major thing that needs to change is the percolate query, but there are some minor things like the string field type which are deprecated and we may as well update at the same time.

I think everything we need to do is technically supported in elasticsearch 5, except that we can't reindex until we switch to the new percolate field. I think it's best to do this in two steps:
- in one PR, add support for the new percolate field and make new indexes with one index per doc type. We don't need to switch to the new indexes right now, but since we need to for the percolate query it would be convenient to do it all at once. We will need to detect which ones we use and have code to handle both cases, to provide for zero downtime.
- the deploy for that PR should use Elasticsearch 5. Elasticsearch will support that upgrade from 2 to 5. Then we should run recreate_index
- next PR should remove all the code handling the old indices.

**Comment by pdpinch** [comment]
What do you need from @mitodl/devops, and at which step do you need it?

**Comment by noisecapella** [comment]
After we merge the PR supporting both ES2 and ES5 we would need to upgrade Elasticsearch. I'm not sure how that's done for them, locally if I switch to a docker container with Elasticsearch 5 it performs the update automatically. Afterwards they would run `./manage.py recreate_index` and with the new PR we would switch over to the new schema. I think that should do it for @mitodl/devops

**Comment by noisecapella** [comment]
@pdpinch I added a deployment step to my PR. In short there are three steps: merge the PR, upgrade Elasticsearch from 2 to 5, run recreate_index.
